File: Node.java
package org.smf;

/**
 * This class represents a node in a simple linked list (the nodes of which store integer values).
 */
public class Node {
    /**
     * The value to be stored in the nod
     */
    int value;

    /**
     * A pointer to the next nod
     */
    Node next;

    /**
     * The constructor builds one node from an integer
     * @param value
     */
    public Node(int value){
        this.value=value;
        next=null;
    }

    /**
     * Builds a string representation of the node (format: [e]).
     */
    public String toString(){
        return "["+value+"]";
    }

    /**
     * This methods appends a node at the end of the linked list.
     * @param n: the node to append
     * @return: the head of the list
     */
    public Node append(Node n){
        Node current=this;
        while (current.next!=null) current=current.next;
        current.next=n;
        return this;
    }

    /**
     * This method implements the visitor design pattern. It accepts the visitor for the current node, then propagate this behavior
     * to the next node.
     * @param v: the visitor of the nod
     */
    public void accept(Visitor v){
        v.visit(this);
        if (next!=null) next.accept(v);
    }

    public static void test1(){
        Node n=new Node(5);
        Node p=new Node(6);
        Node q=new Node(7);
        Node r=new Node(8);
        n.append(p).append(q).append(r);
        n.accept(new DisplayVisitor());
        LengthVisitor l=new LengthVisitor();
        n.accept(l);
        System.out.println(l.getCounter());
    }

    public static void test2(){
        Node n=new Node(5);
        Node p=new Node(6);
        Node q=new Node(7);
        Node r=new Node(8);
        n.append(n).append(q).append(r);
        n.accept(new DisplayVisitor());
        LengthVisitor l=new LengthVisitor();
        n.accept(l);
        System.out.println(l.getCounter());
    }

    public static void test3(){
        Node n=new Node(5);
        Node p=new Node(6);
        Node q=new Node(7);
        Node r=new Node(8);
        n.append(p).append(q).append(r);
        p.next=n;
        n.accept(new DisplayVisitor());
        LengthVisitor l=new LengthVisitor();
        n.accept(l);
        System.out.println(l.getCounter());
    }

    public static void main(String...args){
        test3();
    }
}

File: Node.py
from __future__ import annotations

class Node:
    """This class represents a node in a simple linked list
    """
    
    def __init__(self,value:int) -> None:
        """Build one node which has no next node.

        Args:
            value (int): the value to be stored in the nod
        """
        self.value=value
        self.next=None
        
    def __str__(self) -> str:
        """Builds a string representation of the node

        Returns:
            str: the string representation
        """
        return f"[{self.value}]"
    
    
    def append(self,n:Node) -> Node:
        """Appends a node at the end of the list.

        Args:
            n (Node): the node be added.

        Returns:
            Nod: The head of the list
        """
        if self.next==None:
            self.next=n
        else:
            self.next.append(n)
        return self


    def accept(self,v:Visitor) -> None:
        """This methods implements the visitor design pattern

        Args:
            v (Visitor): the visitor of the node.
        """
        v.visit(self)
        if self.next:self.next.accept(v)

class Visitor:
    """A class that represents a generic visitor
    """
    def visit(self,n:Node) -> None :
        """The behavior of the visitor at node n.

        Args:
            n (Node): the node which is currently visited
        """
        pass

class DisplayVisitor(Visitor):
    """This the visitor which display all the list nodes
    """
    def visit(self, n: Node) -> None:
        end="->" if n.next else "\n"
        print(n,end=end)

class LengthVisitor(Visitor):
    """This the visitor which counts the nodes in the list.
    """
    
    def __init__(self) -> None:
        """A counter is used to count the nodes, initialized to 0.
        """
        self.counter=0
    
    def visit(self, n: Node) -> None:
        self.counter+=1

def test1():
    n=Node(5)
    p=Node(6)
    q=Node(7)
    r=Node(8)
    n.append(p).append(q).append(r)
    n.accept(DisplayVisitor())
    l=LengthVisitor()
    n.accept(l)
    print(l.counter)

def test2():
    n=Node(5)
    p=Node(6)
    q=Node(7)
    r=Node(8)
    n.append(n).append(q).append(r)
    n.accept(DisplayVisitor())
    l=LengthVisitor()
    n.accept(l)
    print(l.counter)

def test3():
    n=Node(5)
    p=Node(6)
    q=Node(7)
    r=Node(8)
    n.append(p).append(q).append(r)
    p.next=n
    n.accept(DisplayVisitor())
    l=LengthVisitor()
    n.accept(l)
    print(l.counter)


if __name__=="__main__":
    test3()
